File: src/pocketcloud/cloud/network/packet/impl/normal/PlayerDisconnectPacket.php
<?php

namespace pocketcloud\cloud\network\packet\impl\normal;

use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\client\ServerClientCache;
use pocketcloud\cloud\network\packet\CloudPacket;
use pocketcloud\cloud\network\packet\data\PacketData;
use pocketcloud\cloud\player\CloudPlayerManager;

final class PlayerDisconnectPacket extends CloudPacket {

    public function __construct(private ?string $playerName = "") {}

    public function encodePayload(PacketData $packetData): void {
        $packetData->write($this->playerName);
    }

    public function decodePayload(PacketData $packetData): void {
        $this->playerName = $packetData->readString();
    }

    public function getPlayer(): string {
        return $this->playerName;
    }

    public function handle(ServerClient $client): void {
        if (($player = CloudPlayerManager::getInstance()->get($this->playerName)) !== null) {
            if ($player->getCurrentProxy() === null) {
                CloudPlayerManager::getInstance()->remove($player);
            } else {
                if (($server = ServerClientCache::getInstance()->getServer($client)) !== null) {
                    if ($server->getTemplate()->getTemplateType()->isProxy()) {
                        CloudPlayerManager::getInstance()->remove($player);
                    }
                }
            }
        }
    }

    public static function create(string $player): self {
        return new self($player);
    }
}

File: cloud-network/src/main/java/de/pocketcloud/network/packet/impl/PlayerDisconnectPacket.java
package de.pocketcloud.network.packet.impl;

import de.pocketcloud.api.network.packet.AuthenticatedPacket;
import de.pocketcloud.api.network.packet.CloudboundPacket;
import de.pocketcloud.api.network.packet.data.IPacketData;
import de.pocketcloud.network.packet.CloudPacket;
import lombok.Getter;
import lombok.NoArgsConstructor;

@NoArgsConstructor
@Getter
public final class PlayerDisconnectPacket extends CloudPacket implements CloudboundPacket, AuthenticatedPacket {

    private String player;

    public PlayerDisconnectPacket(String player) {
        this.player = player;
    }

    @Override
    public void encodePayload(IPacketData packetData) {}

    @Override
    public void decodePayload(IPacketData packetData) {
        this.player = packetData.readString();
    }

    public static PlayerDisconnectPacket create(String player) {
        return new PlayerDisconnectPacket(player);
    }
}
